### PR TITLE
Do not compute payload sha256 with https endpoint

### LINKED ISSUE #428: Disable sha256 computation for https

### PR DESCRIPTION
**Pull request by vadmeste**
Fixes #428 

It would be better to make the layer responsible for putting objects chooses what hash algorithms should be computed

### CODE REVIEW

File: api-put-object-multipart.go [hunk 1]
```diff
@@ -114,7 +117,15 @@ func (c Client) putObjectMultipartStream(bucketName, objectName string, reader i
...
-		md5Sum, sha256Sum, prtSize, rErr := c.hashCopyN(tmpBuffer, reader, partSize)
+		hashSums := make(map[string][]byte)
+		hashAlgos := make(map[string]hash.Hash)
+
+		hashAlgos["md5"] = md5.New()
+		if c.signature.isV4() && !c.secure {
+			hashAlgos["sha256"] = sha256.New()
+		}
+
+		prtSize, rErr := hashCopyN(hashAlgos, hashSums, tmpBuffer, reader, partSize)
```
Comment (krishnasrinivas) [inline api-put-object-multipart.go]: with this approach, hashCopyN and related functions (hashCopy hashCopyBuffer) can be eliminated by doing:

```
md5Writer := md5.New()
writers := []io.Writers{tmpBuffer, md5Writer}

var sha256Writer hash.Hash

if c.signature.isV4() && !c.secure {
    sha256Writer = sha256.New()
    writers = append(writers, sha256Writer)
}

wn, err := io.CopyN(io.MultiWriter(writers...), reader, partSize)
```

File: api.go [hunk 2]
```diff
@@ -583,9 +588,13 @@ func (c Client) newRequest(method string, metadata requestMetadata) (req *http.R
...
-			req.Header.Set("X-Amz-Content-Sha256", hex.EncodeToString(sum256([]byte{})))
-			if metadata.contentSHA256Bytes != nil {
-				req.Header.Set("X-Amz-Content-Sha256", hex.EncodeToString(metadata.contentSHA256Bytes))
+			if !c.secure {
+				req.Header.Set("X-Amz-Content-Sha256", hex.EncodeToString(sum256([]byte{})))
+				if metadata.contentSHA256Bytes != nil {
+					req.Header.Set("X-Amz-Content-Sha256", hex.EncodeToString(metadata.contentSHA256Bytes))
+				}
+			} else {
+				req.Header.Set("X-Amz-Content-Sha256", "UNSIGNED-PAYLOAD")
```
Comment (krishnasrinivas) [inline api.go]: better flow would be:
```
shaHeader := "UNSIGNED-PAYLOAD"
if !c.secure {
    if metadata.contentSHA256Bytes == nil {
        shaHeader = hex.EncodeToString(sum256([]byte{})))
    } else {
        shaHeader = hex.EncodeToString(metadata.contentSHA256Bytes)
    }
}
req.Header.Set("X-Amz-Content-Sha256", shaHeader)
```

Comment (vadmeste) [inline api.go]: :+1:

File: api-put-object-multipart.go (hunk 1 above)
Comment (vadmeste) [inline api-put-object-multipart.go]: > with this approach, hashCopyN and related functions (hashCopy hashCopyBuffer) can be eliminated

@krishnasrinivas, actually, hashCopyN is called from three places, I guess we have to keep it factored

Comment (krishnasrinivas) [inline api-put-object-multipart.go]: >> with this approach, hashCopyN and related functions (hashCopy hashCopyBuffer) can be eliminated

> @krishnasrinivas, actually, hashCopyN is called from three places, I guess we have to keep it factored

similar change has to be done at other places also, our own implementation of hash* functions are not providing advantage anymore. So why not use golang io package's utility functions io.Copy io.CopyN io.CopyBuffer.

File: api-put-object-file.go [hunk 3]
```diff
@@ -177,9 +180,16 @@ func (c Client) putObjectMultipartFromFile(bucketName, objectName string, fileRe
 		sectionReader := io.NewSectionReader(fileReader, totalUploadedSize, partSize)
 
 		// Calculates MD5 and SHA256 sum for a section reader.
-		var md5Sum, sha256Sum []byte
+		hashAlgos := make(map[string]hash.Hash)
+		hashSums := make(map[string][]byte)
+
+		hashAlgos["md5"] = md5.New()
```
Comment (harshavardhana) [inline api-put-object-file.go]: Needs a comment here, why this is needed.

File: api-put-object-multipart.go [hunk 4]
```diff
@@ -114,7 +117,15 @@ func (c Client) putObjectMultipartStream(bucketName, objectName string, reader i
 	for partNumber <= totalPartsCount {
 		// Calculates MD5 and SHA256 sum while copying partSize bytes
 		// into tmpBuffer.
-		md5Sum, sha256Sum, prtSize, rErr := c.hashCopyN(tmpBuffer, reader, partSize)
+		hashSums := make(map[string][]byte)
+		hashAlgos := make(map[string]hash.Hash)
+
+		hashAlgos["md5"] = md5.New()
```
Comment (harshavardhana) [inline api-put-object-multipart.go]: Needs a comment as well.

File: api-put-object.go [hunk 5]
```diff
@@ -193,11 +196,19 @@ func (c Client) putObjectSingle(bucketName, objectName string, reader io.Reader,
 	if size <= -1 {
 		size = maxSinglePutObjectSize
 	}
-	var md5Sum, sha256Sum []byte
+	// var md5Sum, sha256Sum []byte
```
Comment (harshavardhana) [inline api-put-object.go]: This comment should be changed and fixed to reflect the following changes..

File: api.go [hunk 6]
```diff
@@ -163,6 +166,8 @@ func privateNew(endpoint, accessKeyID, secretAccessKey string, secure bool) (*Cl
 		clnt.anonymous = true
 	}
 
+	clnt.secure = secure
```
Comment (harshavardhana) [inline api.go]: Comment?

**Pull request merged**